### ISSUE TITLE
Race conditions and app hangs when power cycling with SPC debugger open

### ISSUE THREAD

**Issue by nyanpasu64** [NovaSquirrel/Mesen-SX#58]
Mono 6.12.0 on Arch Linux (not native Windows), Mesen-SX 2a275e253d39a3b5a026b3bbec708e8b36d79b6b.

- Load a SPC file so it plays.
- Open SPC debugger. (paused)
- Focus main window.
- Press Ctrl+T 3 times fast.
- If Mesen-S doesn't hang, close and reopen the emulator and try again.

The app will hang, looping half a second of audio. The gdb stack trace says:

```
#0  0x00007ff817cd5345 in SimpleLock::AcquireSafe() () at /usr/lib/libMesenSCore.dll
#1  0x00007ff817df5052 in Console::GetDebugger(bool) () at /usr/lib/libMesenSCore.dll
#2  0x00007ff817c50176 in GetDisassemblyLineData () at /usr/lib/libMesenSCore.dll
#3  0x0000000041241b95 in  ()
#4  0x0000000000000000 in  ()
```

If you spam menu items such as pausing and power-cycling, sometimes a power cycle will leave the emulator running. At this point, if you pause the emulator from the SPC Debugger's toolbar, Mesen-S hangs and blocks all mouse input (keyboard works) to all apps in the entire X11 session, which is worse than the app alone hanging. This stack trace is different:

```
#0  0x00007fffdbdf5f4f in Console::IsThreadPaused() () at /usr/lib/libMesenSCore.dll
#1  0x00007fffdbded588 in DebugBreakHelper::DebugBreakHelper(Debugger*) () at /usr/lib/libMesenSCore.dll
#2  0x00007fffdbe1b8fd in Debugger::Step(CpuType, int, StepType) () at /usr/lib/libMesenSCore.dll
#3  0x00007fffdbc4ff27 in Step () at /usr/lib/libMesenSCore.dll
#4  0x000000004045abe9 in  ()
#5  0x00007fffbbfb8530 in  ()
...
```

`Console::IsThreadPaused` returns quickly, but `DebugBreakHelper::DebugBreakHelper` loops forever (`finish` never returns).

**Comment by nyanpasu64** [comment]
I managed to build Mesen-SX using Mono's csc/xbuild.

## Reproducing the hang

I found a code change to deterministically hang the debugger:

https://github.com/NovaSquirrel/Mesen-SX/blob/2a275e253d39a3b5a026b3bbec708e8b36d79b6b/UI/Debugger/frmDebugger.cs#L327

`EmuApi.PowerCycle()` is a `[DllImport(DllPath = "MesenSCore.dll")] public static extern void PowerCycle();`, corresponding to InteropDLL/EmuApiWrapper.cpp (I didn't trace this code further yet):

https://github.com/NovaSquirrel/Mesen-SX/blob/2a275e253d39a3b5a026b3bbec708e8b36d79b6b/InteropDLL/EmuApiWrapper.cpp#L232-L237

I edited this code to call `EmuApi.PowerCycle()` twice, then ran the code in gdb using:

```
make -j12 && gdb -args mono bin/x64/Release/Mesen-S.exe
handle SIG35 noprint nostop
handle SIG36 noprint nostop
r
```

(If I omit noprint from both handlers, the bug does not occur. I think this means a race condition prevented by GDB prints serializing execution.)

 If I open the SPC debugger, unpause, and in the debugger menu Debug > Power Cycle, the emulator hangs in the same spot as before:

```
(gdb) bt
#0  0x00007fffdbcd5345 in SimpleLock::AcquireSafe() () at /home/nyanpasu64/code/Mesen-SX/bin/x64/Release/libMesenSCore.dll
#1  0x00007fffdbdf5052 in Console::GetDebugger(bool) () at /home/nyanpasu64/code/Mesen-SX/bin/x64/Release/libMesenSCore.dll
#2  0x00007fffdbc50176 in GetDisassemblyLineData () at /home/nyanpasu64/code/Mesen-SX/bin/x64/Release/libMesenSCore.dll
#3  0x000000004044c3e5 in  ()
#4  0x0000000000000000 in  ()
```

[`Console::GetDebugger`](https://github.com/NovaSquirrel/Mesen-SX/blob/2a275e253d39a3b5a026b3bbec708e8b36d79b6b/Core/Console.cpp#L860-L873) is trying to acquire [`Console::_debuggerLock`](https://github.com/NovaSquirrel/Mesen-SX/blob/2a275e253d39a3b5a026b3bbec708e8b36d79b6b/Core/Console.h#L81), even though `Console` has multiple locks. The plural of mutex is deadlock, but I'm not sure that's what's going on here.

More importantly, I don't even know which threads are in play. I could check in gdb I guess (there's discussion of multithreaded gdb backtraces at [pipewire](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/2151#note_1265443)).

The only other reference to `_debuggerLock` is in [`Console::LoadRom()`](https://github.com/NovaSquirrel/Mesen-SX/blob/2a275e253d39a3b5a026b3bbec708e8b36d79b6b/Core/Console.cpp#L387). I should try breaking there.

## Failed lead

Somehow, changing UI/Emulation/ShortcutHandler.cs to power-cycle twice doesn't reproduce the bug, even though making it not power-cycle makes the power cycle menu item a no-op:

https://github.com/NovaSquirrel/Mesen-SX/blob/2a275e253d39a3b5a026b3bbec708e8b36d79b6b/UI/Emulation/ShortcutHandler.cs#L77

**Comment by nyanpasu64** [comment]
I forked Mesen-SX at https://github.com/nyanpasu64/Mesen-SX/tree/log-tid.

I replaced `SimpleLock` with a wrapper around `std::recursive_mutex`. Unfortunately this didn't fix the bug. I also added mutex logging statements, which made the bug harder to reproduce, but still intermittently reproducible by holding down Ctrl+T. (Another odd behavior is that if spamming Ctrl+T *doesn't* lockup the emulator, it pauses the emulator, and resuming pauses a second time.)

The trace that appears just before the deadlock is:

```
thread 14308 LoadRom locking _debuggerLock 0x55a015169178
thread 14286 GetDebugger locking _debuggerLock 0x55a015169178
(blocks on acquiring, never succeeds)
thread 14308 GetDebugger locking _debuggerLock 0x55a015169178
thread 14308 GetDebugger unlocking _debuggerLock 0x55a015169178
```

Thread 14286 is stuck as usual:

```
#2  0x00007fb3e7cd42ed in SimpleLock::AcquireSafe() () at /home/nyanpasu64/code/Mesen-SX/bin/x64/Release/libMesenSCore.dll
#3  0x00007fb3e7df3d44 in Console::GetDebugger(bool) () at /home/nyanpasu64/code/Mesen-SX/bin/x64/Release/libMesenSCore.dll
#4  0x00007fb3e7c4f176 in GetDisassemblyLineData () at /home/nyanpasu64/code/Mesen-SX/bin/x64/Release/libMesenSCore.dll
```

For some reason, thread 14308 never returns from LoadRom and releases _debuggerLock. Debugging in gdb, thread 14308 is stuck in Mono internals:

```
#0  0x00007fb40708215a in __futex_abstimed_wait_common () at /usr/lib/libc.so.6
#1  0x00007fb407084960 in pthread_cond_wait@@GLIBC_2.3.2 () at /usr/lib/libc.so.6
#2  0x000055a0127e358e in  ()
#3  0x000055a01273a5dd in  ()
#4  0x000055a01273bdcb in  ()
#5  0x000055a01273c8a0 in  ()
#6  0x000055a01272cf0b in  ()
#7  0x000055a0126ba8e4 in  ()
#8  0x0000000040e68592 in  ()
#9  0x00007fb3a40a79f0 in  ()
#10 0x00007fb406bd1278 in  ()
#11 0x00007fb4068dfea0 in  ()
#12 0x0000000000000000 in  ()
```

In `layout asm` and `f 4`, scrolling up a bit I see:

```
  0x55a01273a5c3  addr32 call 0x55a0127f1ca0 <mono_threads_enter_gc_safe_region_internal>
  0x55a01273a5c9  mov    %r12d,%edx
  0x55a01273a5cc  lea    0x10(%rbp),%rsi
  0x55a01273a5d0  lea    0x38(%rbp),%rdi
  0x55a01273a5d4  mov    %rax,%r13
  0x55a01273a5d7  addr32 call 0x55a0127e34a0
> 0x55a01273a5dd  mov    %r14,%rsi
  0x55a01273a5e0  mov    %r13,%rdi
  0x55a01273a5e3  mov    %eax,%r12d
  0x55a01273a5e6  addr32 call 0x55a0127f1d80 <mono_threads_exit_gc_safe_region_internal>
```

I think there's a deadlock between the emulation thread 14308 acquiring `_debuggerLock`, entering a GC, and waiting on a condvar that the GUI thread(?) never notifies, and the main/GUI thread 14286 blocking forever acquiring `_debuggerLock`, and never returning from native code to Mono to enter a GC safepoint/safe region, to allow the emulation thread's GC to proceed and release `_debuggerLock`.

To me, it's extremely suspect that the emulation thread even has GC pauses in the first place. In any case, this bug is likely heavily platform-dependent, and I don't know if it happens on real .NET on Windows, or if there's any easy way to fix this (eg. avoid GC pauses in the emulation thread). I don't know how Mono works.

I still don't know how `LoadRom` vanished off the call stack of Thread 14308 without returning, or why Thread 14308 has three stack frames in the 0x00007fb40... region, near but not in libMesenSCore.dll.

**Comment by nyanpasu64** [comment]
I managed to get a Pernosco trace (https://pernos.co/debug/ZOzCDNWsgZyKtXJWlr_bTQ/index.html), which gave me some more debugging information.

-   [emulation thread 471682](https://pernos.co/debug/ZOzCDNWsgZyKtXJWlr_bTQ/index.html#f{m[A5tY,AA_,t[GA,BzKC_,f{e[Axmt,Kqg_,s{aJqdnUGAA,bAmw,uVSKA,oWG5g___,v[{wiLg,v[{f'list',q'stdouterr',p{_,xAYag_,{f'container',q'stack',p{_,xAYag___,{w/eg,v[{f'source',q'source',p{_,xAYag_,{f'container',q'current-tasks',p{_,xAYag____/):
	-   Console::LoadRom
		-   acquires _debuggerLock
		-   Enters a GC without returning from LoadRom
			- on Pernosco (unlike gdb), LoadRom doesn't vanish from the stack trace. See below for how `Console::LoadRom()` calls into Mono.
			- The GC is waiting on a "safe point" condvar that the GUI thread(?) never notifies
-   [main/GUI thread 471659](https://pernos.co/debug/ZOzCDNWsgZyKtXJWlr_bTQ/index.html#f{m[A5tY,AA_,t[AQ,BzJr_,f{e[AxnL,ARPu_,s{aJqdnUGAA,bAmw,uVSKA,oWPm6___,v[{wiLg,v[{f'list',q'stdouterr',p{_,xAYag_,{f'container',q'stack',p{_,xAYag___,{w/eg,v[{f'source',q'source',p{_,xAYag_,{f'container',q'current-tasks',p{_,xAYag____/):
	-   GetDisassemblyLineData -> GetDebugger -> Console::GetDebugger(bool autoStart = true)
		-   blocks forever acquiring _debuggerLock
		-   never returns from native code to Mono to enter a GC safepoint/safe region, to allow the emulation thread's GC to proceed and release _debuggerLock.

The stack trace of the emulation thread during the deadlock (most recent call last):

```
Console::LoadRom (this=0x101630318cf0, romFile={_path="/home/nyanpasu64/code/AddMusicKFF/SPCs/volume-broken.spc", _innerFile="", _innerFileIndex=-1, _data=""}, patchFile={_path="", _innerFile="", _innerFileIndex=-1, _data=""}, stopRom=false, forPowerCycle=true) at Console.cpp:462
NotificationManager::SendNotification (this=0x5c0766dbfe30, type=GameLoaded, parameter=0x1) at NotificationManager.cpp:47
InteropNotificationListener::ProcessNotification (this=0x24fb27b769b0, type=GameLoaded, parameter=0x1) at InteropNotificationListener.h:24
0x6f3733235040+136
…
__GI___pthread_cond_wait
```

**Comment by nyanpasu64** [comment]
In [Console::LoadRom](https://github.com/NovaSquirrel/Mesen-SX/blob/fd4b1233da090d6f40d174939912d7885f9d1abb/Core/Console.cpp#L397=), is it safe to move `SendNotification(ConsoleNotificationType::GameLoaded...)` to the very end of `if(cart)`? Or to unlock the mutex when performing this call, then relock it afterwards?

**Comment by nyanpasu64** [comment]
The bug also happens on real Windows and .NET.

- Load a SPC file so it plays.
- Open SPC debugger. (paused)
- Focus main window.
- Press Ctrl+T 3 times fast.

![Discord_6osSlxvdzK](https://user-images.githubusercontent.com/913957/173458971-dacabbf0-db0a-455c-ba1f-24072a8ea8ac.png)

**Comment by nyanpasu64** [comment]
Reading https://blog.ffwll.ch/2022/08/locking-hierarchy.html reminds me of this issue:

> cross-release dependency between the entire `work_fn()` and `cancel_work_sync()`

`stop_processing()` enters `cancel_work_sync()` and waits for `work_fn()` to return, before releasing the lock. And `work_fn()` waits for `stop_processing()` to release the lock, before returning.

Similarly in Mesen-SX, `Console::LoadRom()` enters GC and waits for `Console::GetDebugger()` to return, before releasing the lock. And `Console::GetDebugger()` waits for `Console::LoadRom()` to release the lock, before returning.

> Like with interrupt handlers the clean solution tends to be an additional nested lock which protects just the mutable state shared with the work function and nests within the main object lock. That way work can be cancelled while the main object lock is held, which avoids a ton of races. But without holding the sublock that `work_fn()` needs, which avoids the deadlock.

Is there some way for `Console::GetDebugger()` to lock only a part of the state, without incurring data inconsistencies or race conditions from insufficient mutual exclusion? It's been months since I looked at the code, so I'm not sure what to do here.

> Note that in some cases the superior lock doesn't need to exist, e.g. `struct drm_connector_state` is protected by the [single owner pattern](https://blog.ffwll.ch/2022/08/locking-hierarchy.html#locking-pattern-single-owner), but drivers might have some need for some further decoupled asynchronous processing, e.g. for handling the content protect or link training machinery. In that case only the sublock for the mutable driver private state shared with the worker exists.

IDK if that applies here.

I haven't read https://lwn.net/Articles/709849/ yet.

**Comment by nyanpasu64** [comment]
Idea: unlock the debugger lock after the `if(_debugger) { ... }` block ends? Either call an unlock method on the lock, or add a scope so the lock goes out of scope afterwards. This is safe *if* the lock exists to protect the `_debugger` pointer only.
